File: 剑指 Offer（第 2 版）/16.cpp
// 剑指 Offer 16. 数值的整数次方
// https://leetcode-cn.com/problems/shu-zhi-de-zheng-shu-ci-fang-lcof/

/**
 * 当 x=0 时，直接返回0（避免1/x的影响）
 * 当 n<0 时，要把问题转化为 n>=0 的范围内，x=1/x，n=-n
 * 二分法快速幂：n=0时跳出循环
 *    1. n%2 == 1   ans*=x
 *    2. x*=x
 *    3. n/=2
 */

using namespace std;

class Solution {
public:
    double myPow(double x, long long n) {
        if(!x){
            return 0;
        }
        double ans = 1;
        if(n<0){
            x = 1/x;
            n = -n;
        }
        while(n){
            if(n&1 == 1){
                ans *= x;
            }
            x *= x;
            n >>= 1;
        }
        return ans;
    }
};

File: 剑指 Offer（第 2 版）/16.java
class Solution {
    public double myPow(double x, int n) {
        if(x==0) return 0;
        // 要用long，不然−2147483648会报错
        /* Java 代码中 int32 变量 n 的范围 [-2147483648, 2147483647]
        因此当 n = -2147483648 时执行 n = -n 会因越界而赋值出错。
        解决方法是先将 n 存入 long 变量 b ，后面用 b 操作即可。*/
        long b = n;
        if(b<0) {
            x = 1/x;
            b = -b;
        }
        double res = 1.0;
        while(b>0){
            if(b%2==1) res *= x;
            x *= x;
            b >>= 1;
        }
        return res;
    }
}